Add python 3.11 to the CI

bdraco opened the pull request:
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add python 3.11 to the CI
Previous PR https://github.com/home-assistant/core/pull/74650

### Breaking enum changes
some breaking changes to Enum https://blog.pecar.me/python-enum
- [x] https://github.com/home-assistant/core/pull/88087

### cchardet has a fork now:  https://pypi.org/project/faust-cchardet/
- [x] vulcan-api change https://github.com/kapi2289/vulcan-api/pull/126
- [ ] docker https://github.com/home-assistant/docker/pull/260
- [ ] bump docker
- [x] bump lib https://github.com/home-assistant/core/pull/88254
- [x] ban abandoned cchardet


### Remaining known issues (this list will probably grow)

#### InterruptibleThreadPoolExecutor needs to be updated for py3.11

- [x] investigate - this was a small timing issue, its behaving as expected

#### Issues with pending fixes
- [x] cloud https://github.com/home-assistant/core/pull/88088
- [x] group https://github.com/home-assistant/core/pull/88089
- [x] hassio https://github.com/home-assistant/core/pull/88090
- [x] forked_daapd https://github.com/home-assistant/core/pull/88091
- [x] mqtt https://github.com/home-assistant/core/pull/88092
- [x] netatmo https://github.com/home-assistant/core/pull/88093
- [x] rmvtransport https://github.com/home-assistant/core/pull/88094
- [x] sleepiq https://github.com/home-assistant/core/pull/88098
- [x] unifiprotect  https://github.com/home-assistant/core/pull/88101 
- [x] pyunifiprotect https://github.com/home-assistant/core/pull/88130
- [x] deconz https://github.com/home-assistant/core/pull/88114
- [x] twentemilieu  https://github.com/home-assistant/core/pull/88121
- [x] zwave_me https://github.com/home-assistant/core/pull/88122
- [x] wemo https://github.com/home-assistant/core/pull/88123
- [x] vera https://github.com/home-assistant/core/pull/88124
- [x] arcam https://github.com/home-assistant/core/pull/88125
- [x] apprise https://github.com/home-assistant/core/pull/88126
- [x] enum util https://github.com/home-assistant/core/pull/88131
- [x] yeelight https://github.com/home-assistant/core/pull/88134
- [x] zha https://github.com/home-assistant/core/pull/88144
- [x] Profiler - drop `service.memory` https://github.com/home-assistant/core/pull/88136
- [x] aprs - https://github.com/home-assistant/core/pull/88189
- [x] matter - https://github.com/home-assistant/core/pull/88258

#### Known issues that need to be broken out of here in a separate PR
- [x] <https://github.com/Azure/azure-uamqp-python/issues/334>
   fixed by https://github.com/home-assistant/core/pull/88176
   - aqualink
   - Axis
   - Azure Service Bus
  

#### Integrations using a stub for asyncio.coroutine in tests only (broken in py3.11 but will still run in CI to keep py3.10 support)
- [ ] rflink https://github.com/aequitas/python-rflink/pull/70 - AttributeError: module 'asyncio' has no attribute 'coroutine'
- [x] fido https://github.com/home-assistant/core/pull/88343 https://github.com/titilambert/pyfido/pull/5 - AttributeError: module 'asyncio' has no attribute 'coroutine'
- [ ] blackbird https://github.com/koolsb/pyblackbird/pull/8 - AttributeError: module 'asyncio' has no attribute 'coroutine'




## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

bdraco commented on .github/workflows/ci.yaml:
```diff
@@ -888,7 +888,6 @@ jobs:
             --cov="homeassistant" \
             --cov-report=xml \
             -o console_output_style=count \
-            -p no:sugar \
```
only for debugging

bdraco commented:
Looks like the custom executor need to be adapter as well

bdraco committed:
Remove profiler.memory service

guppy3 is not python3.11 compat
https://github.com/zhuyifei1999/guppy3/issues/41

This service will return if and when guppy3 becomes
python3.11 compat

bdraco committed:
squash

bdraco committed:
Merge branch 'dev' into python3.11

bdraco committed:
temp remove

bdraco committed:
Merge remote-tracking branch 'upstream/python3.11' into python3.11

bdraco committed:
temp dump tests

bdraco committed:
temp dump tests

bdraco committed:
drop a few more to get a run

bdraco committed:
drop a few more to get a run

bdraco committed:
Account for changed python3.11 enum.IntFlag behavior in zha

There may be additional changes needed, but I could only
see what needed to be updated based on the tests

bdraco committed:
Merge branch 'zha_fix_py3.11' into python3.11

bdraco committed:
merge

bdraco commented:
Maybe mock out asyncio.Coro for py3.11 so they still work with 10

bdraco commented:
Still not sure what's going on with the executor since it works locally

bdraco committed:
restore

bdraco committed:
restore

bdraco committed:
legacy value

bdraco committed:
Merge branch 'zha_fix_py3.11' into python3.11

frenck commented:
> ### Not sure what to do about this one
> * [ ]  cchardet has a fork now:  [pypi.org/project/faust-cchardet](https://pypi.org/project/faust-cchardet/)

This is, as far as I spotted, used in two places:

- `vulcan-api` package: <https://github.com/kapi2289/vulcan-api/blob/792a9ec42fd7f059e9e6850cc135dc1bad0ca1c2/setup.py#L38>
  For that one, we could open a PR to replace it.

- Our Docker base images, as it speeds up `aiohttp`. As it seems to be a drop in replacement, we can just replace it. `aiohttp` will try to import it: <https://github.com/aio-libs/aiohttp/blob/3a8e65f3c1cd0f3a22fc7b1ed48d383b453156d5/aiohttp/client_reqrep.py#L72-L75>
  Used here: <https://github.com/home-assistant/docker/blob/c08a196241e3a63792b5f4a6b589a68582d021e6/requirements.txt#LL2-L2C9>

bdraco committed:
Merge remote-tracking branch 'upstream/dev' into python3.11

bdraco committed:
tweak a bit for the python 3.11 timings

bdraco commented:
> * `vulcan-api` package: [kapi2289/vulcan-api@`792a9ec`/setup.py#L38](https://github.com/kapi2289/vulcan-api/blob/792a9ec42fd7f059e9e6850cc135dc1bad0ca1c2/setup.py#L38)
>   For that one, we could open a PR to replace it.

https://github.com/kapi2289/vulcan-api/pull/126

bdraco committed:
block cchardet

bdraco committed:
conditional

bdraco committed:
adjust est

bdraco committed:
adjust est

bdraco committed:
test

bdraco committed:
Merge branch 'profiler' into python3.11

bdraco committed:
not yet

frenck commented:
ticked off the uampq one.

bdraco committed:
tweak

bdraco committed:
Merge branch 'zha_fix_py3.11' into python3.11

bdraco committed:
give a little leeway for timing

emontnemery committed:
Fix otbr tests

bdraco committed:
Merge remote-tracking branch 'upstream/otbr_fix_tests' into python3.11

bdraco committed:
Increase database test timeout

It looks like we need a little more time to run
with the addiitonal tests in #87019

bdraco committed:
Merge branch 'db_timeout' into python3.11

bdraco committed:
Increase database test timeout

It looks like we need a little more time to run
with the addiitonal tests in #87019

bdraco committed:
Merge branch 'db_timeout' into python3.11

bdraco committed:
Fix aprs tests with python 3.11

bdraco committed:
merge fix

bdraco committed:
hints

bdraco committed:
Merge branch 'aprs_py311' into python3.11

bdraco committed:
Merge branch 'dev' into python3.11

bdraco committed:
Merge branch 'dev' into python3.11

bdraco committed:
Update homeassistant/package_constraints.txt

bdraco commented on script/gen_requirements_all.py:
```diff
@@ -150,6 +153,9 @@
 # pyOpenSSL 23.0.0 or later required to avoid import errors when
 # cryptography 39.0.0 is installed with botocore
 pyOpenSSL>=23.0.0
+
+# cmake is required to build uamqp
+cmake==3.25.2
```
```suggestion
```

bdraco committed:
Update script/gen_requirements_all.py

bdraco committed:
Merge branch 'dev' into python3.11

frenck committed:
Constrain uamqp for Python 3.10 only

bdraco committed:
Merge remote-tracking branch 'upstream/dev' into python3.11

bdraco committed:
Merge remote-tracking branch 'upstream/frenck-2023-0312' into python3.11

bdraco committed:
Merge remote-tracking branch 'upstream/python3.11' into python3.11

bdraco committed:
Merge branch 'dev' into python3.11

bdraco committed:
Bump vulcan-api to 2.3.0

see https://github.com/kapi2289/vulcan-api/pull/126
see https://github.com/home-assistant/core/pull/88038
see https://github.com/home-assistant/docker/pull/260

bdraco committed:
Merge branch 'vulcan_api' into python3.11

bdraco committed:
Merge remote-tracking branch 'upstream/python3.11' into python3.11

bdraco committed:
add ban

marcelveldt committed:
Bump python-matter-server to 2.1.1

bdraco committed:
revert

bdraco committed:
Merge remote-tracking branch 'upstream/bump-matter-server-2.1.1' into python3.11

bdraco commented on tests/asyncio_legacy.py:
```diff
@@ -0,0 +1,124 @@
+# flake8: noqa
+# stubbing out for integrations that have
+# not yet been updated for python 3.11
+# but can still run on python 3.10
```
```suggestion
# but can still run on python 3.10
#
# Remove this once rflink, fido, and blackbird
# have had their libraries updated to remove
# asyncio.coroutine
```

bdraco committed:
Update tests/asyncio_legacy.py

bdraco committed:
Merge branch 'dev' into python3.11

bdraco commented:
We can strip out the asyncio test patching at at any point once those integration have their libs updated.

https://github.com/home-assistant/docker/pull/260 will take care of the ccchardet
